lib/MultiTopicsConsumerImpl.cc

```diff
@@ -647,6 +647,41 @@ void MultiTopicsConsumerImpl::acknowledgeAsync(const MessageId& msgId, ResultCal
     }
 }
 
+void MultiTopicsConsumerImpl::acknowledgeAsync(const MessageIdList& messageIdList, ResultCallback callback) {
+    if (state_ != Ready) {
+        callback(ResultAlreadyClosed);
+        return;
+    }
+
+    std::unordered_map<std::string, MessageIdList> topicToMessageId;
+    for (const MessageId& messageId : messageIdList) {
+        auto topicName = messageId.getTopicName();
+        topicToMessageId[topicName].emplace_back(messageId);
+    }
+
+    auto needCallBack = std::make_shared<std::atomic<int>>(topicToMessageId.size());
+    Result res = ResultOk;
+    auto cb = [callback, needCallBack, &res](Result result) {
+        if (result != ResultOk) {
+            res = result;
+        }
+        needCallBack->fetch_sub(1);
+        if (needCallBack->load() == 0) {
+            callback(res);
+        }
```

Comment:
`res` should be defined inside the lambda, not outside. Because it's only used in the lambda.

BTW, please keep it in mind that never capture a reference of a local variable. 
1. The lifetime of the local variable will end after going out of the scope, while the lambda could be invoked in another thread. 
2. Since multiple threads can access the shared local variable, it must be atomic.

That's also why `std::shared_ptr<std::atomic<T>>` is used in other places.
1. Capturing a `std::shared_ptr<U>`  will increase the reference count and extend the lifetime of the `U` instance until all lambdas finished.
2. `U` is `std::atomic<T>`, which is thread safe to access among different threads.

Reply:
You can test this example:

```c++
#include <chrono>
#include <iostream>
#include <thread>
using namespace std;

void f() {
  int x = 0;
  int a[1];  // do nothing
  std::thread t([&x] {
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    cout << std::this_thread::get_id() << " " << &x << " " << x << endl;
  });
  t.detach();
  cout << std::this_thread::get_id() << " " << &x << " " << x << endl;
}

int main() {
  f();
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
}
```

```bash
$ g++ 1.cc -std=c++11 -pthread
$ ./a.out
140118122346304 0x7ffffa28a90c 0
140118122342144 0x7ffffa28a90c 32623
$ ./a.out
139920470566720 0x7fffd23f916c 0
139920470562560 0x7fffd23f916c 32767
```

Reply:
Thank your explanation, I remove `res`. PTAL.